src/helpers/helpers.options.js

```diff
@@ -175,3 +175,64 @@ export function resolve(inputs, context, index, info) {
 		}
 	}
 }
+
+const notIndexable = ['borderDash', 'fill', 'events'];
+const scriptable = (value) => typeof value === 'function';
+const indexable = (key, value) => isArray(value) && notIndexable.indexOf(key) === -1;
+const readKey = (prefix, name) => prefix ? prefix + _capitalize(name) : name;
+const defined = (value) => typeof value !== 'undefined';
+
+/**
+ * Creates a Proxy for resolving raw values for options.
+ * @param {object[]} scopes - The option scopes to look for values, in resolution order
+ * @param {string[]} [prefixes] - The prefixes for values, in resolution order.
+ * @private
+ */
+export function _resolver(scopes, prefixes = ['']) {
+	const cache = Object.create(null);
+	return new Proxy(cache, {
+		get(target, prop) {
+			let value = Reflect.get(target, prop);
+			if (defined(value)) {
+				return value;
+			}
+			let key;
+			loop:
+			for (const prefix of prefixes) {
+				key = readKey(prefix, prop);
+				for (const scope of scopes) {
+					value = scope && scope[key];
+					if (defined(value)) {
+						break loop;
+					}
+				}
+			}
+			if (defined(value)) {
+				Reflect.set(target, prop, value);
+			}
+			return value;
+		}
+	});
+}
+
+/**
+ * Returns an Proxy for resolving option values with context.
+ * @param {Proxy} options - The Proxy returned by `resolver`
+ * @param {object} context - Context object for scriptable/indexable options
+ * @private
+ */
+export function _withContext(options, context) {
```

Comment:
In terms of an API design it might be nice to have a higher level method, that wraps these two and passes the result of `_resolver` to `_withContext`

Something like

```javascript
export function withContext(scope, prefixes, context) {
    const resolver = _resolver(scope, prefixes);
    return _withContext(resolver, context);
}
```

Reply:
`_resolver` = controller/scale level thing, `_withContext` would be per element/tick. (the names are horrible, I know)

Reply:
Ah, so each controller would wrap `_resolver` into something else and the return of that has to be passed to `_withContext`?

Reply:
Yes, the idea is to cache the resolved "raw" values, so we don't need to look up those for each element. And if some of those are scriptable/indexable, only that part needs to be done per element.